test/integration/conftest.py

```diff
@@ -0,0 +1,146 @@
+# Copyright 2025 Canonical Ltd.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# Learn more about testing at: https://juju.is/docs/sdk/testing
+from __future__ import annotations
+
+import functools
+import logging
+import subprocess
+from typing import Callable, Generator
+
+import jubilant
+import minio
+import pytest
+
+
+@pytest.fixture
+def juju() -> Generator[jubilant.Juju, None, None]:
+    """Make a Juju model with the tracing part of COS ready."""
+    with jubilant.temp_model() as j:
+        deploy_tempo(j)
+        deploy_tempo_worker(j)
+        j.deploy('minio', config={'access-key': 'accesskey', 'secret-key': 'mysoverysecretkey'})
+        j.deploy('s3-integrator')
+
+        j.integrate('tempo:s3', 's3-integrator')
+        j.integrate('tempo:tempo-cluster', 'tempo-worker')
+
+        j.wait(lambda s: minio_active(s) and s3_integrator_blocked(s))
+
+        address = j.status().apps['minio'].address
+        mc_client = minio.Minio(
+            f'{address}:9000',
+            access_key='accesskey',
+            secret_key='mysoverysecretkey',
+            secure=False,
+        )
+
+        found = mc_client.bucket_exists('tempo')
+        if not found:
+            mc_client.make_bucket('tempo')
+
+        j.config('s3-integrator', dict(endpoint=f'http://{address}:9000', bucket='tempo'))
+        j.run(
+            's3-integrator/0',
+            'sync-s3-credentials',
+            {'access-key': 'accesskey', 'secret-key': 'mysoverysecretkey'},
+        )
+
+        # Tempo goes through a cycle of:
+        # - update own stateful set
+        # - kill own pod
+        # - new pod is scheduled by k8s
+        # - check own stateful set
+        #
+        # This process may take a while. I'm unsure about the default timeout.
+
+        j.wait(jubilant.all_active)
+
+        yield j
+
+        print(j.debug_log())
+
+
+@pytest.fixture(scope='session')
+def build_charm(
+    pytestconfig: pytest.Config,
+) -> Generator[Callable[[], str], None, None]:
+    """Build the test charm and provide the artefact path.
+
+    Starts building the tracing-tester charm early.
+    Call the fixture value to get the built charm file path.
```

Comment:
This is clever! Though perhaps a bit too clever... how much time does it actually save? Won't the first test have to wait for this anyway? Unless it's saving us a lot of time, I'd prefer just `charmcraft pack` in the GitHub actions, or a simple `subprocess.run()` here.

Reply:
Note that charmcraft suffers from cold start on CI, and GitHub runners are slow.

My local cold start `charmcraft pack` result: `Executed in  128.52 secs`

I would expect GHA to take about 3x longer, so 6 minutes?

Meanwhile the total test run is ~22 minutes today, half of it is taken by concierge, and the other 10 minutes by juju deploy / wait, with tempo charm being particularly slow because it restarts itself.

Reply:
But won't the test(s) have to wait until the pack finishes anyway, so it ends up being no better than just running `charmcraft pack` before the tests?

Reply:
charmcraft pack is a session fixture, which means it's executed first
https://docs.pytest.org/en/stable/reference/fixtures.html#fixture-instantiation-order

then the `juju` fixture is run, which spins up the bulk of the model.

then the test is run, which waits for charmcraft to complete (typically it's already complete).